planner/core/rule_partition_processor.go

```diff
@@ -952,7 +952,10 @@ func (s *partitionProcessor) makeUnionAllChildren(ds *DataSource, pi *model.Part
 			if err != nil {
 				return nil, err
 			}
-			children = append(children, &newDataSource)
+			proj := LogicalProjection{Exprs: expression.Column2Exprs(newDataSource.Schema().Columns)}.Init(newDataSource.ctx, newDataSource.SelectBlockOffset())
```

Comment:
Why adding a projection can solve this problem?

Reply:
Because when datasource is converted into table reader, the virtual generated columns get expanded, and more columns that required by upper executor will be returned, resulted in appending more columns than needed during apply. Hence add a projection would solve it.

It also happens to sort, therefore I added it as an additional test case.

However I am not very satisfied with this solution (seems dirty, also has performance drawback), I am working on a new fix, which directly solves the problem by now allowing table reader to return more columns than required.

Reply:
Nitpick, a tricky point is that, this fix relies on the fact that projection elimination is applied before partition processing, otherwise this new projection is possible to be removed. Though this fact seems to be solid in the foreseeable future, it is indeed a risk. I am fine with this fix actually, would your new fix be in the next PR or would you update this PR directly?

Reply:
Hi Kenan, I was trying hard on the new fix, but it seems it will break a lot of things, making it a highly risky and hard fix to accomplish, and I would like a compromise to the current fix (which I am also fine with now). For your nitpick, I aggree it is a foreseeable risk point, do you have any suggestion on how to overcome the risk?